### PR TITLE
Evaluate in hook for symbolic sets

### PR DESCRIPTION
**Pull request by ana-pantilie**
Fixes https://github.com/kframework/kore/issues/1605

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: kore/src/Kore/Builtin/Set.hs [hunk 1]
```diff
@@ -279,12 +283,35 @@ evalIn =
...
+                    ifTrueAddDefinednessCondition result _set
+                bothConcrete = do
+                    _elem <- hoistMaybe $ Builtin.toKey _elem
+                    _set <- expectConcreteBuiltinSet Set.inKey _set
+                    (Builtin.appliedFunction . asExpandedBoolPattern)
+                        (Map.member _elem _set)
+            setSymbolic <|> bothSymbolic <|> bothConcrete
       where
         asExpandedBoolPattern = Bool.asPattern resultSort
+        ifTrueAddDefinednessCondition result setTerm =
```
Comment (ttuegel) [inline kore/src/Kore/Builtin/Set.hs]: Could we name this function something else, perhaps `returnIfTrueAndDefined`? To me, the name sounds like this function always returns a result, but only adds the definedness condition if the result is `True`, which would be wrong for several reasons. I had to read it several times before I understood. :laughing:

File: kore/src/Kore/Builtin/Set.hs [hunk 2]
```diff
@@ -279,12 +283,35 @@ evalIn =
...
+            let condition =
+                    Conditional.fromPredicate
+                    $ makeCeilPredicate resultSort setTerm
+                trueWithCondition =
+                    Builtin.appliedFunction
+                    $ Pattern.andCondition
+                        (asExpandedBoolPattern result)
+                        condition
+             in if result
```
Comment (ttuegel) [inline kore/src/Kore/Builtin/Set.hs]: Could we lift this `if` to a guard on the function definition?

File: kore/test/Test/Kore/Builtin/Set.hs [hunk 3]
```diff
@@ -232,17 +236,56 @@ test_inElement =
...
-                predicate = mkEquals_ patIn patTrue
-            (===) (Test.Bool.asPattern True) =<< evaluateT patIn
-            (===) Pattern.top                =<< evaluateT predicate
+                conditionTerm = mkAnd patTrue (mkCeil_ patElement)
+            actual <- evaluateT patIn
+            expected <- evaluateT conditionTerm
+            actual === expected
         )
+  where
+    -- genKeys :: Gen [TermLike Variable]
```
Comment (virgil-serbanuta) [inline kore/test/Test/Kore/Builtin/Set.hs]: Do you still need this and the comment a few lines below?

Comment (ana-pantilie) [reply kore/test/Test/Kore/Builtin/Set.hs]: These changes are outdated

**Pull request merged**